src/python/pants/binaries/execution_environment_mixin.py

```diff
@@ -0,0 +1,58 @@
+# coding=utf-8
+# Copyright 2018 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import (absolute_import, division, generators, nested_scopes, print_function,
+                        unicode_literals, with_statement)
+
+import os
+from contextlib import contextmanager
+
+from abc import abstractmethod
+
+from pants.util.contextutil import environment_as
+
+
+def prepend_path(prev_env, path_var, new_entries):
+  prev_path = prev_env.get(path_var, None)
+  if prev_path is None:
+    return ':'.join(new_entries)
+  return ':'.join(new_entries + prev_path.split(':'))
+
+
+class ExecutionEnvironmentMixin(object):
+
+  @classmethod
+  def apply_successive_env_modifications(cls, env, environment_mixin_instances):
+    for env_modifier in environment_mixin_instances:
+      env = env_modifier.modify_environment(env)
+    return env
+
+  # There's no reason this has to be just about the shell "environment", really
+  # just has to be a contextmanager. Implementing this method using a literal
+  # chroot, or VM image, or something might be really interesting to just
+  # completely sidestep the installation problem.
+  @contextmanager
+  def execution_environment(self, prev_env=None):
```

Comment:
It's not clear that this needs to be a member of a mixin. Would it be more complicated as a static method next to `environment_as`... ie, `environment_path_as`, with an argument for the additional path entries?

Reply:
The idea was to abstract over subsystems that provide path entries, because NativeToolchain pulls in either [`XCodeCLITools`](https://github.com/cosmicexplorer/pants/blob/0b1c385cba373b3da4ce2a431f1868c6ede30086/src/python/pants/backend/native/subsystems/platform_specific/darwin/xcode_cli_tools.py) or `Binutils`, depending on the platform, and the same code should consume either one. A mixin seems appropriate for that -- but having the mixin handle the environment and everything seems way overboard. I've changed this to [an extremely simple mixin named `ExecutablePathProvider`](https://github.com/cosmicexplorer/pants/blob/0b1c385cba373b3da4ce2a431f1868c6ede30086/src/python/pants/binaries/binary_tool.py#L146) instead, and consumers such as BuildLocalPythonDistributions can [do any environment modifications themselves](https://github.com/cosmicexplorer/pants/blob/0b1c385cba373b3da4ce2a431f1868c6ede30086/src/python/pants/backend/python/tasks/build_local_python_distributions.py#L116) (I've also [moved the path modification utility method to contextutil](https://github.com/cosmicexplorer/pants/blob/0b1c385cba373b3da4ce2a431f1868c6ede30086/src/python/pants/util/contextutil.py#L30)). Let me know if that addresses your concern, or if I should be abstracting in some other way.